packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,559 @@ const double _kHandleSize = 22.0;
+    }
+  }
+
+  // Returns true when the child should be painted, false otherwise.
+  bool _shouldPaintChild(RenderObject renderObjectChild, int index) {
+    // Paint the navButton when there is overflow.
+    if (renderObjectChild == firstChild) {
+      return _lastIndexThatFits != -1;
+    }
+
+    // If there is no overflow, all children besides the navButton are painted.
+    if (_lastIndexThatFits == -1) {
+      return true;
+    }
+
+    // When there is overflow, paint if the child is in the part of the menu
+    // that is currently open. Overflowing children are painted when the
+    // overflow menu is open, and the children that fit are painted when the
+    // overflow menu is closed.
+    return (index > _lastIndexThatFits) == overflowOpen;
+  }
+
+  // Decide which children will be pained and set their shouldPaint, and set the
+  // offset that painted children will be placed at.
+  void _placeChildren() {
+    int i = -1;
+    Size nextSize = const Size(0.0, 0.0);
+    double fitWidth = 0.0;
+    final RenderBox navButton = firstChild;
+    double overflowHeight = overflowOpen && !isAbove ? navButton.size.height : 0.0;
+    visitChildren((RenderObject renderObjectChild) {
+      i++;
+
+      final RenderBox child = renderObjectChild as RenderBox;
+      final _ToolbarParentData childParentData = child.parentData as _ToolbarParentData;
+
+      // The navigation button is placed after iterating all children, and there
+      // is no need to place children that won't be painted.
+      if (renderObjectChild == firstChild
+          || !_shouldPaintChild(renderObjectChild, i)) {
+        childParentData.shouldPaint = false;
+        return;
+      }
+      childParentData.shouldPaint = true;
+
+      if (!overflowOpen) {
+        childParentData.offset = Offset(fitWidth, 0.0);
+        fitWidth += child.size.width;
+        nextSize = Size(
+          fitWidth,
+          math.max(child.size.height, nextSize.height),
+        );
+      } else {
+        childParentData.offset = Offset(0.0, overflowHeight);
+        overflowHeight += child.size.height;
+        nextSize = Size(
+          math.max(child.size.width, nextSize.width),
+          overflowHeight,
+        );
+      }
+    });
+
+    // Place the navigation button if needed.
+    final _ToolbarParentData navButtonParentData = navButton.parentData as _ToolbarParentData;
+    if (_shouldPaintChild(firstChild, 0)) {
+      navButtonParentData.shouldPaint = true;
+      if (overflowOpen) {
+        navButtonParentData.offset = isAbove
+          ? Offset(0.0, overflowHeight)
+          : Offset.zero;
+        nextSize = Size(
+          nextSize.width,
+          isAbove ? nextSize.height + navButton.size.height : nextSize.height,
+        );
+      } else {
+        navButtonParentData.offset = Offset(fitWidth, 0.0);
+        nextSize = Size(nextSize.width + navButton.size.width, nextSize.height);
+      }
+    } else {
+      navButtonParentData.shouldPaint = false;
```

Comment:
Didn't you already do this above in line 435? Maybe change line 435 to just be:

```
if (renderObjectChild == firstChild) {
  // Deal with the navigation button later.
  return;
}

... and then figure out what to do with the nav button right here.

Reply:
Will do.  I agree this way is misleading now that I look at it.